src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -1,70 +1,113 @@
-            ILookaheadBlockPuller lookaheadBlockPuller,
+            IBlockPuller blockPuller,
             IConsensusRules consensusRules,
-            IFinalizedBlockHeight finalizedBlockHeight)
+            IFinalizedBlockHeight finalizedBlockHeight,
+            IConnectionManager connectionManager,
+            IBlockStore blockStore = null)
         {
             this.network = network;
             this.chainState = chainState;
+            this.blockValidator = blockValidator;
             this.consensusSettings = consensusSettings;
-            this.concurrentChain = concurrentChain;
-            this.lookaheadBlockPuller = lookaheadBlockPuller;
+            this.blockPuller = blockPuller;
             this.consensusRules = consensusRules;
+            this.connectionManager = connectionManager;
+            this.blockStore = blockStore;
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
 
-            this.chainedHeaderTree = new ChainedHeaderTree(network, loggerFactory, chainedHeaderValidator, checkpoints, chainState, finalizedBlockHeight, consensusSettings);
+            this.chainedHeaderTree = new ChainedHeaderTree(network, loggerFactory, blockValidator, checkpoints, chainState, finalizedBlockHeight, consensusSettings);
+
+            this.treeLock = new object();
+            this.reorgLock = new object();
+            this.blockRequestedLock = new object();
 
-            this.blocksRequested = new Dictionary<uint256, List<Action>>();
-            this.toDownloadQueue = new AsyncQueue<(ChainedHeader headerFrom, ChainedHeader headerTo)>();
+            this.blocksRequested = new Dictionary<uint256, List<OnBlockDownloadedCallback>>();
+            this.toDownloadQueue = new Queue<BlockDownloadRequest>();
         }
 
-        public async Task InitializeAsync()
+        /// <summary>
+        /// Set the tip of <see cref="ConsensusManager"/> to be the header that is stored to disk,
+        /// if the given <see cref="consensusTip"/> is not in store then rewind store till a tip is found.
+        /// If <see cref="IChainState.BlockStoreTip"/> is not null (block store is enabled) ensure the store and consensus manager are on the same tip,
+        /// if consensus manager is ahead then reorg it to be the same height as store.
+        /// </summary>
+        /// <param name="consensusTip">The consensus tip that is attempted to be set.</param>
+        public async Task InitializeAsync(ChainedHeader consensusTip)
         {
-            this.logger.LogTrace("()");
+            this.logger.LogTrace("({0}:'{1}')", nameof(consensusTip), consensusTip);
+
+            // TODO: consensus store
+            // We should consider creating a consensus store class that will internally contain
+            //  coinview and it will abstract the methods `RewindAsync()` `GetBlockHashAsync()` 
 
             uint256 utxoHash = await this.consensusRules.GetBlockHashAsync().ConfigureAwait(false);
```

Comment:
I agree that loading block hash from consensus rules looks weird.

Reply:
And could we please find a better name for this variable. How about coin view tip hash?

Reply:
Well I dont want to mention coinview in anyway its a store implementation detail,
I am thinking of creating a ConsensusStore class that internally will have coinview this will look much better.
But we can do it after CM is finished.

Reply:
If you don't want to call it coin view, try something else, but better name should be suggested. How about persisted consensus tip hash?